magda-authorization-api/src/getWhoAllowDatasetOperation.ts

```diff
@@ -0,0 +1,199 @@
+            .split(".");
+        // --- convert camelCase field name to snakecase field name for SQL indentifier
+        sqlIdentiderParts[sqlIdentiderParts.length - 1] = _.snakeCase(
+            sqlIdentiderParts[sqlIdentiderParts.length - 1]
+        );
+
+        const sqlIdentider = sqlIdentiderParts.join(".");
+        if (!isValidSqlIdentifier(sqlIdentider)) {
+            throw new Error(`Invalid Sql Identifier ${sqlIdentider} produced.`);
+        }
+        return sqlIdentider;
+    },
+    "input.user.permissions[_].operations[_]": (ref: string): string => {
+        const sqlIdentiderParts = ref
+            .replace("input.user.permissions[_].operations[_]", "operations")
+            .split(".");
+        // --- convert camelCase field name to snakecase field name for SQL indentifier
+        sqlIdentiderParts[sqlIdentiderParts.length - 1] = _.snakeCase(
+            sqlIdentiderParts[sqlIdentiderParts.length - 1]
+        );
+
+        const sqlIdentider = sqlIdentiderParts.join(".");
+        if (!isValidSqlIdentifier(sqlIdentider)) {
+            throw new Error(`Invalid Sql Identifier ${sqlIdentider} produced.`);
+        }
+        return sqlIdentider;
+    },
+    "input.user.managingOrgUnitIds[_]": (ref: string): string => {
+        if (ref !== "input.user.managingOrgUnitIds[_]") {
+            throw new Error(`Invalid ref returned from OPA: ${ref}`);
+        }
+        return `(SELECT Children."id"
+        FROM org_units AS Parents, org_units AS Children
+        WHERE Children."left" >= Parents."left" AND Children."left" <= Parents."right" AND Parents."id" = users."orgUnitId")`;
+    },
+    "input.user.id": "users.id",
+    "input.user.permissions[_].resourceUri": "resources.uri"
+};
+
+function isValidRef(ref: string): boolean {
+    return (
+        Object.keys(VALID_REF_PREFIX_LIST).findIndex(
+            item => ref.indexOf(item) === 0
+        ) !== -1
+    );
+}
+
+function regoRefToSqlIdentifier(ref: RegoRef): string {
+    const refString = ref.fullRefString();
+    if (!isValidRef(refString)) {
+        throw new Error(`Invalid ref: ${ref.fullRefString()}`);
+    }
+    const prefixes = Object.keys(VALID_REF_PREFIX_LIST).sort(
+        (a: string, b: string) => b.length - a.length
+    );
+    for (let i = 0; i < prefixes.length; i++) {
+        const prefix = prefixes[i];
+        if (!prefix) continue;
+        const idx = refString.indexOf(prefix);
+        if (idx !== 0) continue;
+        if (typeof VALID_REF_PREFIX_LIST[prefix] === "function") {
+            return (VALID_REF_PREFIX_LIST[prefix] as Function)(refString);
+        } else if (typeof VALID_REF_PREFIX_LIST[prefix] === "string") {
+            return VALID_REF_PREFIX_LIST[prefix] as string;
+        } else {
+            throw new Error("Invalid VALID_REF_PREFIX_LIST item type!");
+        }
+    }
+    return refString;
+}
+
+function regoExpToSql(exp: RegoExp, sqlParameters: any[]): string {
```

Comment:
regExp

Reply:
😄 this one really should be RegoExp = Rego Expression for short~
Opa named the language Rego~

Reply:
oh sorry 😂